File: Tetris/Python/blocks.py
import pyray as pr
from position import *
from helpers import colors


# ----------------------------------------------------------------------------------------------------------------------


class Block:
    def __init__(self) -> None:
        """Parent block for other blocks. Should not be used itself."""
        self.id_code = 0
        self.positions: dict[int, list[Position]] = {}
        self.cell_size = 30
        self.rotation = 0
        self.colors: list[pr.Color] = colors()
        self.offset_rows = 0
        self.offset_columns = 0
        self.move(0, 3)

    # -------------------------------------------------------------------------

    def draw(self, offset_x: int, offset_y: int) -> None:
        """Draws the block onto the grid."""

        for item in self.get_cell_positions():
            pr.draw_rectangle(
                item.column * self.cell_size + offset_x,
                item.row * self.cell_size + offset_y,
                self.cell_size - 1,
                self.cell_size - 1,
                self.colors[self.id_code],
            )

    # -------------------------------------------------------------------------

    def move(self, rows: int, columns: int) -> None:
        """Moves the block around the grid."""

        self.offset_rows += rows
        self.offset_columns += columns

    # -------------------------------------------------------------------------

    def get_cell_positions(self) -> list[Position]:
        """Calculates the tile positions considering the current offsets."""

        current_positions: list[Position] = self.positions[self.rotation]
        moved_tiles: list[Position] = []

        for tile in current_positions:
            new_position = Position(
                tile.row + self.offset_rows, tile.column + self.offset_columns
            )
            moved_tiles.append(new_position)

        return moved_tiles

    # -------------------------------------------------------------------------

    def rotate_block(self) -> None:
        """Changes the block to its next state."""

        self.rotation += 1
        if self.rotation > len(self.positions) - 1:
            self.rotation = 0

    # -------------------------------------------------------------------------

    def undo_rotation(self) -> None:
        """Reverts to the previous rotation."""

        self.rotation -= 1
        if self.rotation < 0:
            self.rotation = len(self.positions) - 1


# ----------------------------------------------------------------------------------------------------------------------


class LBlock(Block):
    def __init__(self) -> None:
        super().__init__()
        self.id_code = 1
        self.fill_positions()

    # -------------------------------------------------------------------------

    def fill_positions(self) -> None:
        """Defines the position of all the block's states."""

        self.positions[0] = [
            Position(0, 2),
            Position(1, 0),
            Position(1, 1),
            Position(1, 2),
        ]
        self.positions[1] = [
            Position(0, 1),
            Position(1, 1),
            Position(2, 1),
            Position(2, 2),
        ]
        self.positions[2] = [
            Position(1, 0),
            Position(1, 1),
            Position(1, 2),
            Position(2, 0),
        ]
        self.positions[3] = [
            Position(0, 0),
            Position(0, 1),
            Position(1, 1),
            Position(2, 1),
        ]


# ----------------------------------------------------------------------------------------------------------------------


class JBlock(Block):
    def __init__(self) -> None:
        super().__init__()
        self.id_code = 2
        self.fill_positions()

    # -------------------------------------------------------------------------

    def fill_positions(self) -> None:
        """Defines the position of all the block's states."""

        self.positions[0] = [
            Position(0, 0),
            Position(1, 0),
            Position(1, 1),
            Position(1, 2),
        ]
        self.positions[1] = [
            Position(0, 1),
            Position(0, 2),
            Position(1, 1),
            Position(2, 1),
        ]
        self.positions[2] = [
            Position(1, 0),
            Position(1, 1),
            Position(1, 2),
            Position(2, 2),
        ]
        self.positions[3] = [
            Position(0, 1),
            Position(1, 1),
            Position(2, 0),
            Position(2, 1),
        ]


# ----------------------------------------------------------------------------------------------------------------------


class IBlock(Block):
    def __init__(self) -> None:
        super().__init__()
        self.id_code = 3
        self.fill_positions()
        self.move(-1, 0)

    # -------------------------------------------------------------------------

    def fill_positions(self) -> None:
        """Defines the position of all the block's states."""

        self.positions[0] = [
            Position(1, 0),
            Position(1, 1),
            Position(1, 2),
            Position(1, 3),
        ]
        self.positions[1] = [
            Position(0, 2),
            Position(1, 2),
            Position(2, 2),
            Position(3, 2),
        ]
        self.positions[2] = [
            Position(2, 0),
            Position(2, 1),
            Position(2, 2),
            Position(2, 3),
        ]
        self.positions[3] = [
            Position(0, 1),
            Position(1, 1),
            Position(2, 1),
            Position(3, 1),
        ]


# ----------------------------------------------------------------------------------------------------------------------


class OBlock(Block):
    def __init__(self) -> None:
        super().__init__()
        self.id_code = 4
        self.fill_positions()
        self.move(0, 1)

    # -------------------------------------------------------------------------

    def fill_positions(self) -> None:
        """Defines the position of all the block's states."""

        self.positions[0] = [
            Position(0, 0),
            Position(0, 1),
            Position(1, 0),
            Position(1, 1),
        ]


# ----------------------------------------------------------------------------------------------------------------------


class SBlock(Block):
    def __init__(self) -> None:
        super().__init__()
        self.id_code = 5
        self.fill_positions()

    # -------------------------------------------------------------------------

    def fill_positions(self) -> None:
        """Defines the position of all the block's states."""

        self.positions[0] = [
            Position(0, 1),
            Position(0, 2),
            Position(1, 0),
            Position(1, 1),
        ]
        self.positions[1] = [
            Position(0, 1),
            Position(1, 1),
            Position(1, 2),
            Position(2, 2),
        ]
        self.positions[2] = [
            Position(1, 1),
            Position(1, 2),
            Position(2, 0),
            Position(2, 1),
        ]
        self.positions[3] = [
            Position(0, 0),
            Position(1, 0),
            Position(1, 1),
            Position(2, 1),
        ]


# ----------------------------------------------------------------------------------------------------------------------


class TBlock(Block):
    def __init__(self) -> None:
        super().__init__()
        self.id_code = 6
        self.fill_positions()

    # -------------------------------------------------------------------------

    def fill_positions(self) -> None:
        """Defines the position of all the block's states."""

        self.positions[0] = [
            Position(0, 1),
            Position(1, 0),
            Position(1, 1),
            Position(1, 2),
        ]
        self.positions[1] = [
            Position(0, 1),
            Position(1, 1),
            Position(1, 2),
            Position(2, 1),
        ]
        self.positions[2] = [
            Position(1, 0),
            Position(1, 1),
            Position(1, 2),
            Position(2, 1),
        ]
        self.positions[3] = [
            Position(0, 1),
            Position(1, 0),
            Position(1, 1),
            Position(2, 1),
        ]


# ----------------------------------------------------------------------------------------------------------------------


class ZBlock(Block):
    def __init__(self) -> None:
        super().__init__()
        self.id_code = 7
        self.fill_positions()

    # -------------------------------------------------------------------------

    def fill_positions(self) -> None:
        """Defines the position of all the block's states."""

        self.positions[0] = [
            Position(0, 0),
            Position(0, 1),
            Position(1, 1),
            Position(1, 2),
        ]
        self.positions[1] = [
            Position(0, 2),
            Position(1, 1),
            Position(1, 2),
            Position(2, 1),
        ]
        self.positions[2] = [
            Position(1, 0),
            Position(1, 1),
            Position(2, 1),
            Position(2, 2),
        ]
        self.positions[3] = [
            Position(0, 1),
            Position(1, 0),
            Position(1, 1),
            Position(2, 0),
        ]


# ----------------------------------------------------------------------------------------------------------------------

File: Tetris/Golang/blocks.go
package main

// Index 0 = block type
// Index 1 = rotation
// Index 2 = cells
// Index 3 = x or y

type block [][][]int32

type Blocks struct {
	blocks []block
}

// ---------------------------------------------------------------------------------------------------------------------

func (b *Blocks) init() {
	b.lBlockPositions()
	b.jBlockPositions()
	b.iBlockPositions()
	b.oBlockPositions()
	b.sBlockPositions()
	b.tBlockPositions()
	b.zBlockPositions()
}

// ---------------------------------------------------------------------------------------------------------------------

func (b *Blocks) lBlockPositions() {
	var lBlock block
	lBlock = append(lBlock, [][]int32{{0, 2}, {1, 0}, {1, 1}, {1, 2}})
	lBlock = append(lBlock, [][]int32{{0, 1}, {1, 1}, {2, 1}, {2, 2}})
	lBlock = append(lBlock, [][]int32{{1, 0}, {2, 0}, {1, 1}, {1, 2}})
	lBlock = append(lBlock, [][]int32{{0, 0}, {0, 1}, {1, 1}, {2, 1}})

	b.blocks = append(b.blocks, lBlock)
}

// ---------------------------------------------------------------------------------------------------------------------

func (b *Blocks) jBlockPositions() {
	var jBlock block
	jBlock = append(jBlock, [][]int32{{0, 0}, {1, 0}, {1, 1}, {1, 2}})
	jBlock = append(jBlock, [][]int32{{0, 1}, {0, 2}, {1, 1}, {2, 1}})
	jBlock = append(jBlock, [][]int32{{1, 0}, {1, 1}, {1, 2}, {2, 2}})
	jBlock = append(jBlock, [][]int32{{0, 1}, {1, 1}, {2, 0}, {2, 1}})

	b.blocks = append(b.blocks, jBlock)
}

// ---------------------------------------------------------------------------------------------------------------------

func (b *Blocks) iBlockPositions() {
	var iBlock block
	iBlock = append(iBlock, [][]int32{{1, 0}, {1, 1}, {1, 2}, {1, 3}})
	iBlock = append(iBlock, [][]int32{{0, 2}, {1, 2}, {2, 2}, {3, 2}})
	iBlock = append(iBlock, [][]int32{{1, 0}, {1, 1}, {1, 2}, {1, 3}})
	iBlock = append(iBlock, [][]int32{{0, 2}, {1, 2}, {2, 2}, {3, 2}})

	b.blocks = append(b.blocks, iBlock)
}

// ---------------------------------------------------------------------------------------------------------------------

func (b *Blocks) oBlockPositions() {
	var oBlock block
	oBlock = append(oBlock, [][]int32{{0, 0}, {0, 1}, {1, 0}, {1, 1}})
	oBlock = append(oBlock, [][]int32{{0, 0}, {0, 1}, {1, 0}, {1, 1}})
	oBlock = append(oBlock, [][]int32{{0, 0}, {0, 1}, {1, 0}, {1, 1}})
	oBlock = append(oBlock, [][]int32{{0, 0}, {0, 1}, {1, 0}, {1, 1}})

	b.blocks = append(b.blocks, oBlock)
}

// ---------------------------------------------------------------------------------------------------------------------

func (b *Blocks) sBlockPositions() {
	var sBlock block
	sBlock = append(sBlock, [][]int32{{0, 1}, {0, 2}, {1, 0}, {1, 1}})
	sBlock = append(sBlock, [][]int32{{0, 1}, {1, 1}, {1, 2}, {2, 2}})
	sBlock = append(sBlock, [][]int32{{1, 1}, {1, 2}, {2, 0}, {2, 1}})
	sBlock = append(sBlock, [][]int32{{0, 0}, {1, 0}, {1, 1}, {2, 1}})

	b.blocks = append(b.blocks, sBlock)
}

// ---------------------------------------------------------------------------------------------------------------------

func (b *Blocks) tBlockPositions() {
	var tBlock block
	tBlock = append(tBlock, [][]int32{{0, 1}, {1, 0}, {1, 1}, {1, 2}})
	tBlock = append(tBlock, [][]int32{{0, 1}, {1, 1}, {1, 2}, {2, 1}})
	tBlock = append(tBlock, [][]int32{{1, 0}, {1, 1}, {1, 2}, {2, 1}})
	tBlock = append(tBlock, [][]int32{{0, 1}, {1, 0}, {1, 1}, {2, 1}})

	b.blocks = append(b.blocks, tBlock)
}

// ---------------------------------------------------------------------------------------------------------------------

func (b *Blocks) zBlockPositions() {
	var zBlock block
	zBlock = append(zBlock, [][]int32{{0, 0}, {0, 1}, {1, 1}, {1, 2}})
	zBlock = append(zBlock, [][]int32{{0, 2}, {1, 1}, {1, 2}, {2, 1}})
	zBlock = append(zBlock, [][]int32{{1, 0}, {1, 1}, {2, 1}, {2, 2}})
	zBlock = append(zBlock, [][]int32{{0, 1}, {1, 0}, {1, 1}, {2, 0}})

	b.blocks = append(b.blocks, zBlock)
}

// ---------------------------------------------------------------------------------------------------------------------
